ntp-client/src/Network/NTP/Client.hs

```diff
@@ -71,248 +68,308 @@ data NtpStatus =
+    = race_
+       ( threadDelay t )
+       ( atomically $ do
+           s <- readTVar tvar
+           check $ s == NtpSyncPending
+       )
+
+data QueryOutcome = Timeout | Result NtpOffset
+
+runQueryLoop ::
+       Tracer IO NtpTrace
+    -> NtpClientSettings
+    -> TVar NtpStatus
+    -> TVar (Maybe [ReceivedPacket])
+    -> These (Socket, [AddrInfo]) (Socket, [AddrInfo])
+    -> IO (Either IOError ())
+runQueryLoop tracer ntpSettings ntpStatus inQueue servers = tryIOError $ forever $ do
+    traceWith tracer NtpTraceClientStartQuery
+    void $ atomically $ writeTVar inQueue $ Just []
+    (_id, outcome) <- runThreads
+    case outcome of
+        Timeout -> do
+            traceWith tracer NtpTraceUpdateStatusQueryFailed
+            atomically $ writeTVar ntpStatus NtpSyncUnavailable
+        Result offset -> do
+             traceWith tracer $ NtpTraceUpdateStatusClockOffset $ getNtpOffset offset
+             atomically $ writeTVar ntpStatus $ NtpDrift offset
+
+    void $ atomically $ writeTVar inQueue Nothing
+    traceWith tracer NtpTraceClientSleeping
+    threadDelayInterruptible ntpStatus $ fromIntegral $ ntpPollDelay ntpSettings
     where
-    -- Receive responses from the network and update NTP client state.
-    loop :: AddrFamily -> Socket.Socket -> IO ()
-    loop addressFamily sock
-        = handle (handleIOException addressFamily) $ forever $ do
-            (bs, _) <- recvFrom sock ntpPacketSize
-            case decodeOrFail $ LBS.fromStrict bs of
-                Left  (_, _, err)    ->
-                    traceWith tracer $ NtpTraceReceiveLoopDecodeError err
-                Right (_, _, packet) ->
-                    handleNtpPacket packet
-
-    -- Restart the @loop@ in case of errors; wait 5s before recreating the
-    -- socket.
-    handleIOException
-        :: AddrFamily
-        -> IOException
+        runThreads
+           = withAsync (sendBoth servers >> loopForever)      $ \sender ->
+             withAsync (timeout      >> return Timeout)   $ \delay ->
+             withAsync (checkReplies >>= return . Result) $ \revc ->
+                    waitAnyCancel [sender, delay, revc]
+
+        timeout = do
+            threadDelay $ fromIntegral $ ntpResponseTimeout ntpSettings
+            traceWith tracer NtpTraceClientWaitingForRepliesTimeout
+
+        checkReplies = atomically $ do
+            q <- readTVar inQueue
+            case q of
+               Nothing -> error "Queue of incomming packets not available"
+               Just l -> case (ntpReportPolicy ntpSettings) l of
+                   Wait -> retry
+                   Report r -> return r
+
+        sendBoth s = case s of
+            This v4 -> send v4
+            That v6 -> send v6
+            These v4 v6 -> send v6 >> send v4
+
+        send (sock, addrs) = forM_ addrs $ \addr -> do
+            p <- mkNtpPacket
+            void $ Socket.ByteString.sendTo sock (LBS.toStrict $ encode p) (setNtpPort $ Socket.addrAddress addr)
```

Comment:
Instead of using expensive `LBS.toStrict` it's better to use vectored I/O in general (e.g. `sendManyTo (LBS.toChunks $ encode p)`.  It should be ok here as each ntp packet easily fits into a single chunk `32kb`, just a comment would do.